lib/runtime/wazero/imports.go

```diff
@@ -0,0 +1,2247 @@
+}
+
+func ext_misc_print_num_version_1(ctx context.Context, m api.Module, data uint64) {
+	logger.Debugf("num: %d", int64(data))
+}
+
+func ext_misc_print_utf8_version_1(ctx context.Context, m api.Module, dataSpan uint64) {
+	data := read(m, dataSpan)
+	logger.Debug("utf8: " + string(data))
+}
+
+// GetRuntimeVersion finds the runtime version by initiating a temporary
+// runtime instance using the WASM code provided, and querying it.
+func GetRuntimeVersion(code []byte) (version runtime.Version, err error) {
+	config := Config{
+		LogLvl: log.DoNotChange,
+	}
+	instance, err := NewInstance(code, config)
+	if err != nil {
+		return version, fmt.Errorf("creating runtime instance: %w", err)
+	}
+	defer instance.Stop()
+
+	version, err = instance.Version()
+	if err != nil {
+		return version, fmt.Errorf("running runtime: %w", err)
+	}
+
+	return version, nil
+}
+
+func ext_misc_runtime_version_version_1(ctx context.Context, m api.Module, dataSpan uint64) uint64 {
+	rtCtx := ctx.Value(runtimeContextKey).(*runtime.Context)
+	if rtCtx == nil {
+		panic("nil runtime context")
+	}
+	code := read(m, dataSpan)
+
+	var option *[]byte
+	version, err := GetRuntimeVersion(code)
+	if err != nil {
+		logger.Errorf("failed to get runtime version: %s", err)
+		ret, err := write(m, rtCtx.Allocator, scale.MustMarshal(option))
+		if err != nil {
+			panic(err)
+		}
+		return ret
+	}
+
+	// Note the encoding contains all the latest Core_version fields as defined in
+	// https://spec.polkadot.network/#defn-rt-core-version
+	// In other words, decoding older version data with missing fields
+	// and then encoding it will result in a longer encoding due to the
+	// extra version fields. This however remains compatible since the
+	// version fields are still encoded in the same order and an older
+	// decoder would succeed with the longer encoding.
+	encodedData, err := scale.Marshal(version)
+	if err != nil {
+		logger.Errorf("failed to encode result: %s", err)
+		return 0
+	}
+
+	option = &encodedData
+
+	ret, err := write(m, rtCtx.Allocator, scale.MustMarshal(option))
+	if err != nil {
+		panic(err)
+	}
+	return ret
+}
+
+func ext_default_child_storage_read_version_1(
+	ctx context.Context, m api.Module, childStorageKey, key, valueOut uint64, offset uint32) uint64 {
+	rtCtx := ctx.Value(runtimeContextKey).(*runtime.Context)
+	if rtCtx == nil {
+		panic("nil runtime context")
+	}
+
+	keyToChild := read(m, childStorageKey)
+	keyBytes := read(m, key)
+	value, err := rtCtx.Storage.GetChildStorage(keyToChild, keyBytes)
+	if err != nil {
+		logger.Errorf("failed to get child storage: %s", err)
+		return 0
+	}
+
+	valueBuf, _ := splitPointerSize(valueOut)
+	ok := m.Memory().Write(valueBuf, value[offset:])
+	if !ok {
+		panic("write overflow")
+	}
+
+	size := uint32(len(value[offset:]))
+	sizeBuf := make([]byte, 4)
+	binary.LittleEndian.PutUint32(sizeBuf, size)
+
+	// this is expected to be Option(Vec<u8>)
+	ret, err := write(m, rtCtx.Allocator, scale.MustMarshal(&sizeBuf))
```

Comment:
Note: previously we did not panic in this (and probably others in this file) case

Reply:
we shouldn't have any buffer overflow errors.  This would be a catastrophic error in this case, so I believe we should panic in this case.

Reply:
makes sense to me